File: src/rogue_example.cpp
#include <rogue/interfaces/stream/Master.h>
#include <rogue/interfaces/stream/Slave.h>
#include <rogue/interfaces/stream/Frame.h>
#include <rogue/interfaces/stream/FrameIterator.h>
#include <rogue/interfaces/stream/Fifo.h>
#include <rogue/interfaces/stream/Buffer.h>
#include <rogue/interfaces/stream/FrameLock.h>
#include <iostream>
#include <vector>

namespace ris = rogue::interfaces::stream;                                     // Creating namespace for stream interface

//Shared pointer aliases
typedef std::shared_ptr<ris::Master> MasterPtr;
typedef std::shared_ptr<ris::Slave> SlavePtr;
typedef std::shared_ptr<ris::Fifo> FifoPtr;


int main(int argc, char const *argv[])
{
    std::cout << "*************** ROGUE CPP EXAMPLES ******************" << "\n\n\n";

    MasterPtr src  = ris::Master::create();                                     // Creation of Master & Slave object ptrs
    SlavePtr  dst  = ris::Slave::create();
    MasterPtr srcf = ris::Master::create();
    SlavePtr  dstf = ris::Slave::create();
    FifoPtr   fifo = ris::Fifo::create(100,0,true);
    dst->setDebug(100,"Primary Slave");                                        // Set debug option for slave. Outputs frame data when sent from master
    dstf->setDebug(100, "Slave");
    ris::FramePtr frame;
    ris::FrameIterator it;
    uint32_t x;
    //uint32_t size;
    //uint64_t data64;
    //uint32_t data32;
    //uint8_t data8;
    //uint8_t * data;

    *src >> dst;                                                               // Connecting Master to Slave

    // Output addresses for Master & Slave Pointers
    std::cout << "Master Pointer Location: ";
    std::cout << src << "\n";
    std::cout << "Slave Pointer Location : ";
    std::cout << dst << "\n";

    std::cout << "Master Slave Count: " << src->slaveCount() << "\n\n";

    // We will request an empty frame from the primary slave (dst)
    frame = src->reqFrame(100,true);                                           // First param is min. size of frame & second param is true for zero copy frame
                                                                               // setting false allows master to send the same frame multiple times
    frame->setPayload(10);                                                     // Set payload size

    it = frame->begin();

    for (x=0; x<10; x++) {
       *it = x*2;
       it++;
    }

    // ********************************************************
    // Another method to fill frame
    // while ( it != frame->end()) {
    //    size = it.remBuffer();                                                  // remBuffer() gets the remaining bytes in the current buffer
       
    //    it = std::copy(data, data+size, it);
    //    data += size;
    // }
    // *******************************************************************

    std::cout << "FRAME ATTRIBUTES *********************************\n";
    std::cout << "Size        : " << frame->getSize() << "\n";
    std::cout << "Available   : " << frame->getAvailable() << "\n";
    std::cout << "Payload     : " << frame->getPayload() << "\n";
    std::cout << "Buffer Count: " << frame->bufferCount() << "\n\n";

    std::cout << "Primary Slave Debug ***************************************************\n";     
    src->sendFrame(frame);                                                     // sendFrame() sends the passed frame to all current slaves
                                                                               // if zeroCopy after sendFrame(arg) returns arg frame will be emptied
    
    std::cout << "\nFRAME SENT **************************************\n\n";
    // std::cout << "Size        : " << frame->getSize() << "\n";
    // std::cout << "Available   : " << frame->getAvailable() << "\n";
    // std::cout << "Payload     : " << frame->getPayload() << "\n";
    // std::cout << "Buffer Count: " << frame->bufferCount() << "\n\n";

    std::cout << "Slave Attributes **********************************\n";
    std::cout << "Frame Count: " << dst->getFrameCount() << "\n";
    std::cout << "Byte Count : " << dst->getByteCount() << "\n\n";

    // ***********************************************************
    // REQUESTING ANOTHER FRAME
    // frame = src->reqFrame(100,true);
    // frame->setPayload(13);

    // it = frame->begin();
   
    // toFrame(it, 8, &data64);                                                 // Following methods should write bit data with byte offsets corresponding
                                                                               // to the second paramater
    // toFrame(it, 4, &data32);
   
    // toFrame(it, 1, &data8);
    // ***********************************************************

    // ***********************************************************
    // FIRST METHOD FOR SENDING FRAME WITH DATA
    // for (x=0; x<10; x++) {
    //    *it = x;
    //    it++;
    // }

    //it = std::copy(data, data+10,it);

    //src->sendFrame(frame);
    // ************************************************************

    // RECEIVING FRAMES
    ris::FrameLockPtr lock = frame->lock();
    std::vector<uint32_t> data(frame->getPayload());
    std::vector<uint32_t>::iterator vit = data.begin();
    
    it = frame->begin();
    // Accessing the frame data using the iterator. For loop to iterate through the iterator and de-reference
    // it to retreive the data.
    // for (x=0; x < 10; x++) {
    //     printf("Location %i = 0x%x\n", x, *it);
    //     it++;
    // }
    // std::cout << "\n\n";
    // std::copy(frame->begin(),frame->end(),std::ostream_iterator<uint32_t>(std::cout<< " "));

    std::copy(frame->begin(),frame->end(), vit);

    for (vit = data.begin(); vit != data.end(); vit++) {
        printf("0x%x ", *vit);
    }
    std::cout<< "\n";

    // ****************************** FIFO DEMONSTRATION *****************************************

    *(*fifo << srcf) >> dstf;                                                    // Connecting the fifo to the source and then onto the destination.


    return 0;
}

File: Python/rogue_example.py
import pyrogue
import rogue.interfaces.stream as pystream

def newLine():
    print('\n')


print('*****************************************')
print('******* TEST USAGE OF ROGUE *************')
print('*****************************************')
newLine()

print('**** CREATE POINTER TO MASTER OBJECT ****')
myMaster = pystream.Master()
print(myMaster)
newLine()

print('************** INITIAL ******************')
print('Slave Count: ' + str(myMaster._slaveCount()))       # Returns the number of slaves connected to master
newLine()

print('***** CREATE POINTER TO SLAVE OBJECT *****')
mySlave = pystream.Slave()
print(mySlave)
newLine()

print('************** INITIAL ******************')
print('Frame Count: ' + str(mySlave.getFrameCount()))      # Returns the number of frames received
print('Byte Count : ' + str(mySlave.getByteCount()))       # Returns the number of bytes received
newLine()

myMaster >> mySlave                                        # Connects master to slave with >> operation
                                                           # Alternatively can connect with pyrogue.streamConnect(myMaster,mySlave)

# Create a few more slave object pointers
slave1 = pystream.Slave()
slave2 = pystream.Slave()
slave3 = pystream.Slave()

myMaster >> slave1
myMaster >> slave2
myMaster >> slave3

print('*************** Slave1 ******************')
print(slave1)
newLine()
print('*************** Slave2 ******************')
print(slave2)
newLine()
print('*************** Slave3 ******************')
print(slave3)
newLine()

print('Slave Count: ' + str(myMaster._slaveCount()))
newLine()

frame = myMaster._reqFrame(100,True)                       # Requesting frame of size 100 with zero-copy enabled

print('*************** Frame *******************')
print('Frame Size     : ' + str(frame.getSize()))          # Total size of frame
print('Remaining Space: ' + str(frame.getAvailable()))     # Remaining space in frame for payload
print('Payload Size   : ' + str(frame.getPayload()))       # Payload size of frame
newLine()

ba1 = bytearray([i for i in range(10)])
frame.write(ba1,0)                                         # Writes the byte array to offset of 0
                                                           # Location 0-9 are filled
ba2 = bytearray([i*2 for i in range(10)])
frame.write(ba2,50)                                        # Writes the new byte array to offset of 50
                                                           # Locations 50-59 are filled
                                                           # Locations 10-49 are not known

myMaster._sendFrame(frame)

print('************** Slave ********************')
print('Frame Count: ' + str(mySlave.getFrameCount()))
print('Byte Count : ' + str(mySlave.getByteCount()))
newLine()

print('********** Byte Array Sent to Slave ***********')
print(ba1 + ba2)
newLine()

print('*************** Frame *******************')
print('Frame Size     : ' + str(frame.getSize()))          # Total size of frame
print('Remaining Space: ' + str(frame.getAvailable()))     # Remaining space in frame for payload
print('Payload Size   : ' + str(frame.getPayload()))       # Payload size of frame
newLine()
with frame.lock():
    size = frame.getPayload()

    fullData = bytearray(size)                             # Create empty array using size of the payload

    frame.read(fullData,0)                                 # Read data from frame starting at an offset of 0
                                                           # and save it into fullData bytearray
    print('************* Full Payload Data ****************')
    print(fullData)
    newLine()

    partialData1 = bytearray(10)                           # Create 2 empty arrays with size of 10  
    partialData2 = bytearray(10) 

    frame.read(partialData1,0)                             # Read data from offset of 0 (reading the payload data passed
                                                           # from original bytearray (ba)) and write into partialData bytearray
    frame.read(partialData2,50)                            # Reading the last 10 from the payload that was written from the 
                                                           # second byte array
print('************ Original Payload Data **************')
print(partialData1 + partialData2)
newLine()


src = pystream.Master()
dst = pystream.Slave()
fifo = pystream.Fifo(0,0,False)                            # Fifo inherits from both the Master and Slave classes
                                                           # so it retains the same public methods as Master and Slave

src >> fifo >> dst                                         # Connect fifo to master and slave to master
                                                           # Fifo will be the primary slave and be the last to
                                                           # receieve data

print('****************** Fifo ************************')
print(fifo)
newLine()

print('src Slave Count : ' + str(src._slaveCount()))
print('fifo Slave Count: ' + str(fifo._slaveCount()))
newLine()

### Testing the following configuration, GOAL IS TO UNDERSTAND WHAT IS HAPPENING HERE WHEN SRC SENDS A FRAME
### WHILE CONNECTED AS SUCH SRC >> FIFO >> DST
ba = bytearray([i for i in range(10)])

frame = src._reqFrame(100,True)
frame.write(ba,50)                                         # We now write the data from the byte array to the frame

src._sendFrame(frame)

print('frame Size   : ' + str(frame.getSize()))
print('frame Payload: ' + str(frame.getPayload()))
newLine()

print('**************** Slave ********************')
print('Frame Count: ' + str(dst.getFrameCount()))
print('Byte Count : ' + str(dst.getByteCount()))
newLine()

print('**************** Fifo ********************')        # Keeping track of Fifo Data, Generally Fifo does not have a frame count
print('Frame Count: ' + str(fifo.getFrameCount()))         # nor a Byte Count since it passes on the frames to the slave.
print('Byte Count : ' + str(fifo.getByteCount()))
print('Fixed Size : ' + str(fifo.getFixedSize()))
print('Alloc Bytes: ' + str(fifo.getAllocBytes()))
print('Alloc Count: ' + str(fifo.getAllocCount()))
newLine()

## We will now create a new situation where Fifo has the zero-copy flag disabled
src = pystream.Master()
dst = pystream.Slave()
fifo = pystream.Fifo(50,20,False)

